Text replacement animations do not run in the correct order and fail to animate over the correct ranges when text comes in over time.

megangardner opened the pull request:
#### 5c25cad683ccc1832ce4f12f283a032a2be9f4ac
<pre>
Text replacement animations do not run in the correct order and fail to animate over the correct ranges when text comes in over time.
<a href="https://bugs.webkit.org/show_bug.cgi?id=276113">https://bugs.webkit.org/show_bug.cgi?id=276113</a>
<a href="https://rdar.apple.com/128292335">rdar://128292335</a>

Reviewed by Aditya Keerthi and Wenson Hsieh.

I was not using the full async animation infrastructure
and the tracking of the different ranges as the different
chunks came in was not correct. These changes should
streamline the animations and correct some not-as-good names
for some of the variables. There are still some frames that
need a bit of polish, but this is a much better working version
than we have had up to this point.

* Source/WebCore/editing/CharacterRange.h:
(WebCore::CharacterRange::operator== const):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::addSourceTextAnimation):
* Source/WebCore/page/writing-tools/WritingToolsController.h:
* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::compositionSessionDidReceiveTextWithReplacementRange):
* Source/WebKit/Shared/TextAnimationType.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::addTextAnimationForAnimationID):
(WebKit::WebPageProxy::callCompletionHandlerForAnimationID):
(WebKit::WebPageProxy::getTextIndicatorForID):
* Source/WebKit/UIProcess/TextAnimationType.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
(-[WKTextAnimationManager addTextAnimationForAnimationID:withData:]):
(-[WKTextAnimationManager textPreviewsForChunk:completion:]):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::addSourceTextAnimation):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
(WebKit::TextAnimationController::unreplacedRangeForSessionWithID const):
(WebKit::TextAnimationController::contextRangeForTextAnimationID const):
(WebKit::TextAnimationController::removeTransparentMarkersForSessionID):
(WebKit::remainingCharacterRange):
(WebKit::TextAnimationController::addInitialTextAnimation):
(WebKit::TextAnimationController::addSourceTextAnimation):
(WebKit::TextAnimationController::addDestinationTextAnimation):
(WebKit::TextAnimationController::updateUnderlyingTextVisibilityForTextAnimationID):
(WebKit::newlyReplacedCharacterRange): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::addTextAnimationForAnimationID):
(WebKit::WebPage::addSourceTextAnimation):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: <a href="https://commits.webkit.org/280763@main">https://commits.webkit.org/280763@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bf31a3bfdb680bdcd10668242db17db0d6465895

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57514 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36842 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9989 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61136 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7959 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59642 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44475 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8147 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46574 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5640 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59544 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34574 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49682 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27438 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31357 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6985 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6962 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53320 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7257 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62815 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1427 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7350 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53832 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1434 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49707 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53951 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12725 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1218 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32671 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33756 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34841 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33502 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/editing/CharacterRange.h:
```diff
@@ -46,6 +46,11 @@ struct CharacterRange {
     CharacterRange() = default;
     constexpr CharacterRange(uint64_t location, uint64_t length);
 
+    bool operator==(const CharacterRange& range) const
```
I think you can use `= default` here.

megangardner replied:
I have no idea how to format that, and searching our codebase isn't helping...

pxlcoder replied:
`bool operator==(const CharacterRange& range) const = default;`

pxlcoder commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -382,24 +382,46 @@
     auto commandState = finished ? WritingToolsCompositionCommand::State::Complete : WritingToolsCompositionCommand::State::InProgress;
 
-    replaceContentsOfRangeInSession(*state, resolvedRange, attributedText, commandState);
+    m_page->chrome().client().addSourceTextAnimation(session.identifier, range, [weakPage = WeakPtr { *m_page }, weakThis = WeakPtr { *this }, state, resolvedRange, movedFragement = WTFMove(fragment), attributedText, commandState, identifier = session.identifier, sessionRange]() mutable {
```
`WritingToolsController` already has a weak reference to page itself, so you don't need `m_page` in the capture list.

pxlcoder replied:
`movedFragement` (typo) can just be called `fragment`.

pxlcoder commented on Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
```diff
@@ -87,6 +87,7 @@ - (instancetype)initWithWebViewImpl:(WebKit::WebViewImpl&)webView
 
 - (void)addTextAnimationForAnimationID:(NSUUID *)uuid withData:(const WebKit::TextAnimationData&)data
 {
+
```
Unnecessary whitespace.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
```diff
@@ -262,13 +309,13 @@
         removeTransparentMarkersForTextAnimationID(uuid);
     else {
-        auto sessionRange = contextRangeForTextAnimationID(uuid);
+        auto animatonRange = contextRangeForTextAnimationID(uuid);
```
```suggestion
        auto animationRange = contextRangeForTextAnimationID(uuid);
```

pxlcoder commented on Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
```diff
@@ -95,19 +96,30 @@ - (void)addTextAnimationForAnimationID:(NSUUID *)uuid withData:(const WebKit::Te
     case WebKit::TextAnimationType::Source:
         effect = adoptNS([PAL::alloc_WTReplaceSourceTextEffectInstance() initWithChunk:chunk.get() effectView:_effectView.get()]);
+        if ([effect respondsToSelector:@selector(setPreCompletion:)]) {
+            effect.get().preCompletion = makeBlockPtr([weakWebView = WeakPtr<WebKit::WebViewImpl>(_webView), nsuuid = uuid] {
```
```suggestion
            effect.get().preCompletion = makeBlockPtr([weakWebView = WeakPtr<WebKit::WebViewImpl>(_webView), uuid = retainPtr(uuid)] {
```

I think you need to retain, since nothing else guarantees the lifetime of `NSUUID *`?

pxlcoder commented on Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
```diff
@@ -95,19 +96,30 @@ - (void)addTextAnimationForAnimationID:(NSUUID *)uuid withData:(const WebKit::Te
+        if ([effect respondsToSelector:@selector(setPreCompletion:)]) {
+            effect.get().preCompletion = makeBlockPtr([weakWebView = WeakPtr<WebKit::WebViewImpl>(_webView), nsuuid = uuid] {
+                auto strongWebView = weakWebView.get();
+                auto wtfuuid = WTF::UUID::fromNSUUID(nsuuid);
```
```suggestion
                auto animationID = WTF::UUID::fromNSUUID(uuid.get());
```

rr-codes commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -382,24 +382,46 @@
+    if (!fragment) {
+        ASSERT_NOT_REACHED();
+        return;
+    }
```
where is `fragment` or `movedFragment` used?

rr-codes commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -382,24 +382,46 @@
+        if (!document) {
+            ASSERT_NOT_REACHED();
+            return;
+        }
```
nit: a newline before the `replaceContentsOfRangeInSession` call would be nice imo

rr-codes commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -382,24 +382,46 @@
+        // FIXME: We won't be setting the selection after every replace, we need a different way to
+        // caluculate this range.
+        auto selectionRange = document->selection().selection().firstRange();
+        if (!selectionRange)
```
this should also have an `ASSERT_NOT_REACHED` before the return

megangardner replied:
We actually run into this more than I would think, and we need to not be calculating this in this manner anyways, so I'm going to leave it for now.

rr-codes commented on Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
```diff
@@ -95,19 +96,30 @@ - (void)addTextAnimationForAnimationID:(NSUUID *)uuid withData:(const WebKit::Te
         break;
     case WebKit::TextAnimationType::Source:
         effect = adoptNS([PAL::alloc_WTReplaceSourceTextEffectInstance() initWithChunk:chunk.get() effectView:_effectView.get()]);
+        if ([effect respondsToSelector:@selector(setPreCompletion:)]) {
```
Assuming this respondsToSelector was originally adding for staging purposes, it can probably be removed now, right?

pxlcoder commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -397,9 +394,29 @@
 
-    replaceContentsOfRangeInSession(*state, resolvedRange, attributedText, commandState);
+    m_page->chrome().client().addSourceTextAnimation(session.identifier, range, [weakThis = WeakPtr { *this }, state, resolvedRange, attributedText, commandState, identifier = session.identifier, sessionRange]() mutable {
+
```
Nit: Would remove this newline.

pxlcoder commented on Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
```diff
@@ -95,21 +95,28 @@ - (void)addTextAnimationForAnimationID:(NSUUID *)uuid withData:(const WebKit::Te
         effect = adoptNS([PAL::alloc_WTReplaceSourceTextEffectInstance() initWithChunk:chunk.get() effectView:_effectView.get()]);
+        effect.get().preCompletion = makeBlockPtr([weakWebView = WeakPtr<WebKit::WebViewImpl>(_webView), uuid = RetainPtr(uuid)] {
+            auto strongWebView = weakWebView.get();
+            auto wtfuuid = WTF::UUID::fromNSUUID(uuid.get());
```
Maybe just `animationID`? `wtf` in the variable is name is weird 😅

pxlcoder commented on Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
```diff
@@ -95,21 +95,28 @@ - (void)addTextAnimationForAnimationID:(NSUUID *)uuid withData:(const WebKit::Te
+            auto strongWebView = weakWebView.get();
+            if (strongWebView)
+                strongWebView->page().updateUnderlyingTextVisibilityForTextAnimationID(remainingID, true);
+            auto wtfuuid = WTF::UUID::fromNSUUID(nsuuid);
```
Same comment as above.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
```diff
@@ -226,28 +259,42 @@
         ASSERT_NOT_REACHED();
         return;
     }
+    // take out the part that has already been replaced
```
```suggestion
    // Take out the part that has already been replaced.
```

If we're keeping this comment...

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
```diff
@@ -226,28 +259,42 @@
-    auto& animationStates = m_activeTextAnimations.ensure(sessionUUID, [&] {
-        return Vector<TextAnimationState> { };
+    m_webPage->addTextAnimationForAnimationID(destinationTextIndicatorUUID, { TextAnimationType::Final, unstyledRangeUUID }, *textIndicatorData, [weakWebPage = WeakPtr { *m_webPage }, sessionUUID]() mutable {
+
```
Nit: Remove newline.

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1216,23 +1216,35 @@ static bool exceedsRenderTreeSizeSizeThreshold(uint64_t thresholdSize, uint64_t
     protectedPageClient()->addTextAnimationForAnimationID(uuid, styleData);
 }
 
+void WebPageProxy::callCompletionHandlerForAnimationID(const WTF::UUID& uuid)
```
If the web process dies, I think `completionHandlerForAnimationID` is going to be left with stale completion handlers.

`WebPageProxy::resetState` should probably clear the map.

pxlcoder commented on Source/WebCore/page/writing-tools/WritingToolsController.mm:
```diff
@@ -397,9 +394,28 @@
-    m_page->chrome().client().addDestinationTextAnimation(session.identifier, adjustedCharacterRange);
+        auto rangeAfterReplace = characterRange(sessionRange, *selectionRange);
+
+        weakThis->m_page->chrome().client().addDestinationTextAnimation(identifier, rangeAfterReplace);
```
`m_page` should probably be null checked.

pxlcoder commented on Source/WebKit/UIProcess/WebPageProxyInternals.h:
```diff
@@ -294,7 +294,8 @@ struct WebPageProxy::Internals final : WebPopupMenuProxy::Client
 #if ENABLE(WRITING_TOOLS)
-    HashMap<WTF::UUID, WebCore::TextIndicatorData> textIndicatorDataForChunk;
+    HashMap<WTF::UUID, WebCore::TextIndicatorData> textIndicatorDataForAnimationID;
+    HashMap<WTF::UUID, CompletionHandler<void()>> completionHandlerForAnimationID;
```
Asked a question to the team about whether this should be `Function` rather than `CompletionHandler`.

pxlcoder replied:
We are leaving as `CompletionHandler` since it's expected to be called in all cases except where the web process is killed.

pxlcoder reviewed (approved):
Please amend the commit message to remove the title next to the radar.

whsieh commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
```diff
@@ -227,27 +260,40 @@
+    m_webPage->addTextAnimationForAnimationID(destinationTextIndicatorUUID, { TextAnimationType::Final, unstyledRangeUUID }, *textIndicatorData, [weakWebPage = WeakPtr { *m_webPage }, sessionUUID]() mutable {
+        if (!weakWebPage)
+            return;
+        WTFLogAlways("<MMG> calling the completion Handler!!! you did it!");
```
(Please remove this debug logging before landing!)

pxlcoder replied:
That was not there in the version of the patch I reviewed!

megangardner replied:
This sneaked in with the stuff I did to make sure iOS was working.

whsieh commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
```diff
@@ -227,27 +260,40 @@
+        if (!weakWebPage)
+            return;
+        WTFLogAlways("<MMG> calling the completion Handler!!! you did it!");
+        weakWebPage->corePage()->chrome().client().addInitialTextAnimation(sessionUUID);
```
It's a bit weird for WebKit (client) code to need to call into the `Chrome`/`ChromeClient` just to propagate back up to the client layer.

Maybe simpler if you just call into `WebPage` directly?

```suggestion
        weakWebPage->addInitialTextAnimation(sessionUUID);
```

whsieh commented on Source/WebKit/WebKitSwift/TextAnimation/TextAnimationManager.swift:
```diff
@@ -114,8 +114,18 @@ extension TextAnimationManager: UITextEffectView.ReplacementTextEffect.Delegate
+                chunkToEffect.removeValue(forKey: animationID)
+
+                guard let delegate = self.delegate else {
+                    Self.logger.debug("Can't obtain Targeted Preview. Missing delegate." )
```
Nit - extra space before the `)`

```suggestion
                    Self.logger.debug("Can't obtain Targeted Preview. Missing delegate.")
```

pxlcoder commented on Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
```diff
@@ -95,21 +95,28 @@ - (void)addTextAnimationForAnimationID:(NSUUID *)uuid withData:(const WebKit::Te
+                strongWebView->page().updateUnderlyingTextVisibilityForTextAnimationID(remainingID, true);
+            auto animationID = WTF::UUID::fromNSUUID(nsuuid);
+            if (animationID)
+                strongWebView->page().callCompletionHandlerForAnimationID(*animationID);
```
```suggestion
        effect.get().completion = makeBlockPtr([weakWebView = WeakPtr<WebKit::WebViewImpl>(_webView), remainingID = data.unanimatedRangeUUID, uuid = RetainPtr(uuid)] {
            auto strongWebView = weakWebView.get();
            if (strongWebView)
                strongWebView->page().updateUnderlyingTextVisibilityForTextAnimationID(remainingID, true);
            auto animationID = WTF::UUID::fromNSUUID(uuid.get());
            if (animationID)
                strongWebView->page().callCompletionHandlerForAnimationID(*animationID);
```

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
```diff
@@ -148,73 +178,78 @@
+    auto replaceCharacterRange = replacingRange;
+
+    if (previouslyReplacedRange) {
+        if ((*previouslyReplacedRange)->range == replacingRange && (*previouslyReplacedRange)->string == string)
```
```suggestion
        if (previouslyReplacedRange->range == replacingRange && previouslyReplacedRange->string == string)
```

pxlcoder replied:
May even be nice to pass in a `ReplacedRangeAndString` as the parameter, so you can directly compare to that, but I won't hold this PR up for that.

megangardner replied:
So, the reason this is (*previouslyReplacedRange)->range and not previouslyReplacedRange->range is because this is a double-wrapped optional. yes, I know it's crazy, I would like to fix it, but this is where we are right now.

pxlcoder commented on Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
```diff
@@ -148,73 +178,78 @@
+        if ((*previouslyReplacedRange)->range == replacingRange && (*previouslyReplacedRange)->string == string)
+            return;
+
+        replaceCharacterRange = remainingCharacterRange(replacingRange, (*previouslyReplacedRange)->range);
```
```suggestion
        replaceCharacterRange = remainingCharacterRange(replacingRange, previouslyReplacedRange->range);
```

megangardner replied:
ditto

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
```diff
@@ -784,40 +807,42 @@ static void checkColor(WebCore::CocoaColor *color, std::optional<ColorExpectatio
         EXPECT_EQ(1UL, contexts.count);
 
-        EXPECT_WK_STREQ(@"Hey do you wanna go see a movie this weekend", contexts.firstObject.attributedText.string);
+        EXPECT_WK_STREQ(@"Hey do you wanna go see a movie this weekend - start", contexts.firstObject.attributedText.string);
```
(for my own understanding: this is just for ease of debugging right?)

megangardner replied:
yes, I like the more 'reasonable' strings, but then I couldn't tell which order they were supposed to come in, so I added suffixes.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
```diff
@@ -1203,8 +1246,8 @@ static void checkColor(WebCore::CocoaColor *color, std::optional<ColorExpectatio
         [[webView writingToolsDelegate] compositionSession:session.get() didReceiveText:attributedText.get() replacementRange:NSMakeRange(0, 19) inContext:contexts.firstObject finished:YES];
-
-        [webView waitForNextPresentationUpdate];
+        TestWebKitAPI::Util::runFor(0.1_s); // this is needed to let the information load at least a bit or we can fail on a javascript issue.
```
```suggestion
        TestWebKitAPI::Util::runFor(0.1_s); // This is needed to let the information load at least a bit or we can fail on a JavaScript issue.
```

pxlcoder replied:
I'm curious what the JS issue is here? Given `waitForValue` already waits for the image to load.

megangardner replied:
Encountered error: Error Domain=WKErrorDomain Code=4 "A JavaScript exception occurred" UserInfo={WKJavaScriptExceptionLineNumber=1, WKJavaScriptExceptionMessage=TypeError: null is not an object (evaluating 'document.querySelector('img').complete'), WKJavaScriptExceptionColumnNumber=30, WKJavaScriptExceptionSourceURL=file:///..../TestWebKitAPI.resources/, NSLocalizedDescription=A JavaScript exception occurred} while evaluating script: document.querySelector('img').complete

pxlcoder replied:
Oh, the `<img>` element wasn't created. It might be nicer to update the `waitForValue` to also wait for that to happen.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WritingTools.mm:
```diff
@@ -1469,10 +1514,12 @@ static void checkColor(WebCore::CocoaColor *color, std::optional<ColorExpectatio
 
         [[webView writingToolsDelegate] writingToolsSession:session.get() didReceiveAction:WTActionCompositionRestart];
 
+
```
Unnecessary change.

pxlcoder commented on Source/WebKit/UIProcess/TextAnimationType.h:
```diff
@@ -37,7 +37,8 @@ enum class TextAnimationType : uint8_t {
 struct TextAnimationData {
     TextAnimationType style;
-    WTF::UUID remainingRangeUUID { WTF::UUID::emptyValue };
+    WTF::UUID unanimatedRangeUUID { WTF::UUID::emptyValue };
```
Is `{ WTF::UUID::emptyValue }` really necessary?

megangardner replied:
TextAnimationType.h:40:35: error: field of type 'WTF::UUID' has private default constructor
   40 |     WTF::UUID unanimatedRangeUUID { };

So, yes.

pxlcoder reviewed (approved):
Slightly curious how long the test suite takes to run now, given all the waiting. 

(I understand it is necessary due to the nature of the animations, just curious)

megangardner committed:
Text replacement animations do not run in the correct order and fail to animate over the correct ranges when text comes in over time.
https://bugs.webkit.org/show_bug.cgi?id=276113
rdar://128292335

Reviewed by Aditya Keerthi and Wenson Hsieh.

I was not using the full async animation infrastructure
and the tracking of the different ranges as the different
chunks came in was not correct. These changes should
streamline the animations and correct some not-as-good names
for some of the variables. There are still some frames that
need a bit of polish, but this is a much better working version
than we have had up to this point.

* Source/WebCore/editing/CharacterRange.h:
(WebCore::CharacterRange::operator== const):
* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::addSourceTextAnimation):
* Source/WebCore/page/writing-tools/WritingToolsController.h:
* Source/WebCore/page/writing-tools/WritingToolsController.mm:
(WebCore::WritingToolsController::compositionSessionDidReceiveTextWithReplacementRange):
* Source/WebKit/Shared/TextAnimationType.serialization.in:
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::addTextAnimationForAnimationID):
(WebKit::WebPageProxy::callCompletionHandlerForAnimationID):
(WebKit::WebPageProxy::getTextIndicatorForID):
* Source/WebKit/UIProcess/TextAnimationType.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxyInternals.h:
* Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
(-[WKTextAnimationManager addTextAnimationForAnimationID:withData:]):
(-[WKTextAnimationManager textPreviewsForChunk:completion:]):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::addSourceTextAnimation):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.h:
* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
(WebKit::TextAnimationController::unreplacedRangeForSessionWithID const):
(WebKit::TextAnimationController::contextRangeForTextAnimationID const):
(WebKit::TextAnimationController::removeTransparentMarkersForSessionID):
(WebKit::remainingCharacterRange):
(WebKit::TextAnimationController::addInitialTextAnimation):
(WebKit::TextAnimationController::addSourceTextAnimation):
(WebKit::TextAnimationController::addDestinationTextAnimation):
(WebKit::TextAnimationController::updateUnderlyingTextVisibilityForTextAnimationID):
(WebKit::newlyReplacedCharacterRange): Deleted.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::addTextAnimationForAnimationID):
(WebKit::WebPage::addSourceTextAnimation):
* Source/WebKit/WebProcess/WebPage/WebPage.h:

Canonical link: https://commits.webkit.org/280763@main
